### ISSUE TITLE
Speechlm2 NaN loss

### ISSUE THREAD

**Issue by AudranBert** [NVIDIA-NeMo/NeMo#15098]
**Describe the bug**

Loss is always NaN (See image below). 
<img width="1494" height="602" alt="Image" src="https://github.com/user-attachments/assets/9de25d5d-ed88-4f2f-8866-25a587ea18c4" />

In nemo 2.4 it works fine : 
<img width="1058" height="608" alt="Image" src="https://github.com/user-attachments/assets/b5ea6ef8-40c9-4d53-ba60-c008c0848a4f" />

Could be a problem with a package or something?

**Steps/Code to reproduce bug**

Config below:
[config_multitask.yaml](https://github.com/user-attachments/files/23652082/config_multitask.yaml)

**Expected behavior**

Train normally

**Environment details**

Pip freeze : 
[requirements_speechlm2.txt](https://github.com/user-attachments/files/23652137/requirements_speechlm2.txt)

Nemo 2.5.2

**Additional context**

We tried on a 
GPU: H100

We also tried on a DGX Spark with the container : nvcr.io/nvidia/nemo:25.09.01 which has nemo 2.5.1 inside.

Sorry again @stevehuang52  for this issue :D

**Comment by AudranBert** [comment]
We have this
```
Error getting class at nemo.collections.asr.modules.transformer.get_nemo_transformer: Located non-class of type 'function' while loading 'nemo.collections.asr.modules.transformer.get_nemo_transformer'
```
at the beginning of the training, maybe it is linked?

**Comment by AudranBert** [comment]
After some testing we found out that when using batch_size=1 instead of batch_tokens everything is fine. But when using batch_size=2, we have NaN again, is there an issue with the padding (I know that the padding code changed compared to 2.4) or something like that in 2.5?

**Comment by stevehuang52** [comment]
Sorry I'm on leave but maybe @pzelasko can help taking a look?

**Comment by AudranBert** [comment]
Sorry to ping you again @pzelasko 
Also I saw that NeMo will be splitted into multiples repository, where will go SpeechLM collection?

**Comment by parthchadha** [comment]
@pzelasko can you please help with this issue? Thanks.

**Comment by stevehuang52** [comment]
> I saw that NeMo will be split into multiples repository, where will go SpeechLM collection?

@AudranBert it will remain in this repo

**Comment by stevehuang52** [comment]
@AudranBert could you please also share the training command you used to launch the job?

**Comment by AudranBert** [comment]
Hi @stevehuang52, here is the bash script we used to launch the training ([salm_train.py ](https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/speechlm2/salm_train.py)) when using slurm. On the DGX we don't use slurm but still use salm_train.py.
```bash
#!/bin/bash
#SBATCH --job-name=test
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --gres=gpu:1
#SBATCH --cpus-per-task=24
#SBATCH --time=02:00:00
#SBATCH --constraint=h100


MASTER_ADDR=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
MASTER_PORT=6000
NODE_RANK=$SLURM_NODEID

DISTRIBUTED_ARGS="
  --nproc_per_node $SLURM_GPUS_ON_NODE \
  --nnodes $SLURM_NNODES \
  --node_rank $NODE_RANK \
  --rdzv_backend c10d \
  --rdzv_endpoint $MASTER_ADDR:$MASTER_PORT \
  --max_restarts 0
"

srun torchrun $DISTRIBUTED_ARGS salm_train.py \
    --config-path=$CONFIG_PATH \
    --config-name=$CONFIG_NAME \
    ++trainer.max_steps=250 \
    trainer.num_nodes=$SLURM_NNODES \
    exp_manager.exp_dir=$EXPERIMENT_FOLDER \
    exp_manager.name=$EXPERIMENT_NAME
```

**Comment by pzelasko** [comment]
I can't think of anything that would have caused this. Can you enable https://docs.pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly and let us know where specifically did NaN first appear?

**Comment by AudranBert** [comment]
@pzelasko The issue is not present (using the DGX Spark) in the container running NeMo 2.4 and not present in the container running 2.6 so the issue must be caused by a package version in the container/requirements in the 2.5.

**Comment by pzelasko** [comment]
Thanks, I think I reproduced it (still don't know why this is happening though). For me, the loss is OK only on step 0, and NaN on step 1 and onwards. Was it the same in your case?

**Comment by AudranBert** [comment]
> Thanks, I think I reproduced it (still don't know why this is happening though). For me, the loss is OK only on step 0, and NaN on step 1 and onwards. Was it the same in your case?

For me, the loss at the first step is already at NaN.

**Comment by AudranBert** [comment]
We now have access to have NeMo 2.6 env on H100 and I face the same problem, but I have the first step where the loss is normal and after that they are all NaN @pzelasko .

**Comment by pzelasko** [comment]
I started debugging it - the NaN first appears in the activations of the first layer of the convolutional subsampling module (self.perception.encoder.pre_encode.conv[0]). The actual issue is likely in the preprocessor. I'll try to find some time today to get this fixed.

**Comment by pzelasko** [comment]
The preprocessor is actually OK. The issue is that the gradients after the first training step are NaN (even though loss is OK), causing parameters to become NaN, which results in NaN activations on forward pass in the second training step. 
The first NaN gradient seems to come from some layer norm (at least in my debugging env). I'll try to pinpoint the cause but it might take me some time.

**Comment by pzelasko** [comment]
@AudranBert what model/models were you training when the NaN issue occurred? I am doing a sweep over multiple LLM backbones and find the issue appears only in specific LLMs, but most of them actually train OK.

**Comment by AudranBert** [comment]
> [@AudranBert](https://github.com/AudranBert) what model/models were you training when the NaN issue occurred? I am doing a sweep over multiple LLM backbones and find the issue appears only in specific LLMs, but most of them actually train OK.

hi @pzelasko ,

We are using :
```
pretrained_llm: meta-llama/Llama-3.2-1B-Instruct
pretrained_asr: nvidia/canary-1b-v2
```
It is the same we used in NeMo 2.4 which was working fine.

**Comment by AudranBert** [comment]
Any updates @pzelasko ?

**Comment by pzelasko** [comment]
I couldn't reproduce the NaN issue with your models so far.

Let's try this:
Can you take the container that is working OK and update NeMo to either 2.6 or `main` (essentially 2.7 pre-release)? If it works, that'd be an issue with some dependency (then we should try to narrow down which one is it).

**Comment by AudranBert** [comment]
> The preprocessor is actually OK. The issue is that the gradients after the first training step are NaN (even though loss is OK), causing parameters to become NaN, which results in NaN activations on forward pass in the second training step. The first NaN gradient seems to come from some layer norm (at least in my debugging env). I'll try to pinpoint the cause but it might take me some time.

Hi thanks for the answer,
This is a different issue that what you have reproduced earlier?

**Comment by pzelasko** [comment]
Yes, the issue I reproduced was specific to Nemotron v2 and v3 LLM models..

**Comment by AudranBert** [comment]
> [@pzelasko](https://github.com/pzelasko) The issue is not present (using the DGX Spark) in the container running NeMo 2.4 and not present in the container running 2.6 so the issue must be caused by a package version in the container/requirements in the 2.5.

Hi again @pzelasko ,
Upgrading nemo in the NeMo 2.4 container worked: no more NaNs
So to sum up:
| Version                    | Container | On H100 |
|----------------------------|:---------:|:-------:|
| 2.4                        | ✅        | ✅      |
| 2.5                        | ❌        | ❌      |
| 2.6                        | ✅        | ❌      |
| 2.6 in NeMo 2.4 container  | ✅        | will try      |

So it must be a package version issue. There are some differences between the 2.6 H100 env we have and the NeMo 2.6 container.
Here are the 2 envs freezes. We are searching for what causes the issue, we noticed that torch and flash attn 2 have different versions.

[h100.txt](https://github.com/user-attachments/files/25131350/h100.txt) VS [container.txt](https://github.com/user-attachments/files/25131352/container.txt)

**Comment by AudranBert** [comment]
> Yes, the issue I reproduced was specific to Nemotron v2 and v3 LLM models..

We are training LLMs with the Nemotron architecture, it means we will face it once we use our models (instead of llama) for training our multimodal model?

**Comment by pzelasko** [comment]
It will be solved soon. I intend to add https://github.com/NVIDIA-NeMo/Automodel as a dependency to replace HF hub version. Automodel is more optimized, recently added Nano V3 support, and confirmed to work without issue.

**Comment by AudranBert** [comment]
> > [@pzelasko](https://github.com/pzelasko) The issue is not present (using the DGX Spark) in the container running NeMo 2.4 and not present in the container running 2.6 so the issue must be caused by a package version in the container/requirements in the 2.5.
> 
> Hi again [@pzelasko](https://github.com/pzelasko) , Upgrading nemo in the NeMo 2.4 container worked: no more NaNs So to sum up:
> Version 	Container 	On H100
> 2.4 	✅ 	✅
> 2.5 	❌ 	❌
> 2.6 	✅ 	❌
> 2.6 in NeMo 2.4 container 	✅ 	✅
> 
> So it must be a package version issue. There are some differences between the 2.6 H100 env we have and the NeMo 2.6 container. Here are the 2 envs freezes. We are searching for what causes the issue, we noticed that torch and flash attn 2 have different versions.
> 
> [h100.txt](https://github.com/user-attachments/files/25131350/h100.txt) VS [container.txt](https://github.com/user-attachments/files/25131352/container.txt)

Hi,
my colleague successfully used NeMo 2.7 (main branch) in the 2.4 environment on H100, and the NaNs have disappeared. Thanks for your help!

**Comment by pzelasko** [comment]
Glad to hear that and sorry for the trouble. I'll try to dissect the dependency list you provided, maybe we can find the culprit.. Let me know if you find anything on your side (understandable if you don't want to think about this ever again though).
